File: apps/mobile/ios/Sources/Models/NavigationState.swift
//
//  NavigationState.swift
//  TchatApp
//
//  Created by Claude on 22/09/2024.
//

import Foundation

/// Core navigation state entity for cross-platform state synchronization
public struct NavigationState: Codable, Equatable, Hashable {

    // MARK: - Properties

    public let id: String
    public let userId: String
    public let sessionId: String
    public let platform: String
    public var navigationStack: [NavigationStackEntry]
    public var currentRoute: String?
    public var previousRoute: String?
    public let timestamp: Date
    public var syncVersion: Int
    public let metadata: NavigationStateMetadata

    // MARK: - Initialization

    public init(
        id: String = UUID().uuidString,
        userId: String,
        sessionId: String,
        platform: String = "ios",
        navigationStack: [NavigationStackEntry] = [],
        currentRoute: String? = nil,
        previousRoute: String? = nil,
        timestamp: Date = Date(),
        syncVersion: Int = 1,
        metadata: NavigationStateMetadata = NavigationStateMetadata()
    ) {
        self.id = id
        self.userId = userId
        self.sessionId = sessionId
        self.platform = platform
        self.navigationStack = navigationStack
        self.currentRoute = currentRoute
        self.previousRoute = previousRoute
        self.timestamp = timestamp
        self.syncVersion = syncVersion
        self.metadata = metadata
    }

    // MARK: - Computed Properties

    /// Get current navigation depth
    public var depth: Int {
        return navigationStack.count
    }

    /// Check if can go back
    public var canGoBack: Bool {
        return navigationStack.count > 1
    }

    /// Get root route
    public var rootRoute: String? {
        return navigationStack.first?.routeId
    }

    /// Get current route parameters
    public var currentParameters: [String: Any] {
        return navigationStack.last?.parameters ?? [:]
    }

    /// Check if state is synchronized
    public var isSynchronized: Bool {
        return metadata.lastSyncTimestamp != nil &&
               Date().timeIntervalSince(metadata.lastSyncTimestamp!) < 30.0 // 30 seconds
    }

    // MARK: - Navigation Operations

    /// Push new route to navigation stack
    public mutating func push(
        routeId: String,
        parameters: [String: Any] = [:],
        timestamp: Date = Date()
    ) {
        let entry = NavigationStackEntry(
            routeId: routeId,
            parameters: parameters,
            timestamp: timestamp,
            transition: .push
        )

        previousRoute = currentRoute
        currentRoute = routeId
        navigationStack.append(entry)
        incrementVersion()
    }

    /// Pop current route from navigation stack
    public mutating func pop() {
        guard navigationStack.count > 1 else { return }

        let poppedEntry = navigationStack.removeLast()
        previousRoute = currentRoute
        currentRoute = navigationStack.last?.routeId

        // Add reverse transition entry for sync
        var reverseEntry = poppedEntry
        reverseEntry.transition = .pop
        reverseEntry.timestamp = Date()

        incrementVersion()
    }

    /// Replace current route
    public mutating func replace(
        routeId: String,
        parameters: [String: Any] = [:],
        timestamp: Date = Date()
    ) {
        guard !navigationStack.isEmpty else {
            push(routeId: routeId, parameters: parameters, timestamp: timestamp)
            return
        }

        let entry = NavigationStackEntry(
            routeId: routeId,
            parameters: parameters,
            timestamp: timestamp,
            transition: .replace
        )

        previousRoute = currentRoute
        currentRoute = routeId
        navigationStack[navigationStack.count - 1] = entry
        incrementVersion()
    }

    /// Reset navigation to root
    public mutating func popToRoot() {
        guard navigationStack.count > 1 else { return }

        let rootEntry = navigationStack.first!
        previousRoute = currentRoute
        currentRoute = rootEntry.routeId
        navigationStack = [rootEntry]
        incrementVersion()
    }

    /// Reset entire navigation state
    public mutating func reset(
        toRoute routeId: String,
        parameters: [String: Any] = [:]
    ) {
        let entry = NavigationStackEntry(
            routeId: routeId,
            parameters: parameters,
            timestamp: Date(),
            transition: .reset
        )

        previousRoute = currentRoute
        currentRoute = routeId
        navigationStack = [entry]
        incrementVersion()
    }

    // MARK: - Sync Operations

    /// Mark state as synchronized
    public mutating func markSynchronized() {
        metadata = NavigationStateMetadata(
            lastSyncTimestamp: Date(),
            syncAttempts: metadata.syncAttempts,
            conflictResolutionStrategy: metadata.conflictResolutionStrategy,
            customData: metadata.customData
        )
    }

    /// Increment sync version
    private mutating func incrementVersion() {
        syncVersion += 1
    }

    /// Create sync request payload
    public func createSyncRequest() -> NavigationStateSyncRequest {
        return NavigationStateSyncRequest(
            userId: userId,
            sessionId: sessionId,
            platform: platform,
            navigationStack: navigationStack,
            timestamp: Date(),
            syncVersion: syncVersion
        )
    }

    /// Apply sync response
    public mutating func applySyncResponse(_ response: NavigationStateSyncResponse) {
        if response.success {
            self.syncVersion = response.syncVersion
            markSynchronized()
        }
    }
}

// MARK: - Supporting Types

/// Navigation stack entry with transition information
public struct NavigationStackEntry: Codable, Equatable, Hashable, Identifiable {
    public let id: String
    public let routeId: String
    public let parameters: [String: Any]
    public let timestamp: Date
    public var transition: NavigationTransition

    public init(
        id: String = UUID().uuidString,
        routeId: String,
        parameters: [String: Any] = [:],
        timestamp: Date = Date(),
        transition: NavigationTransition = .push
    ) {
        self.id = id
        self.routeId = routeId
        self.parameters = parameters
        self.timestamp = timestamp
        self.transition = transition
    }

    // MARK: - Codable Conformance

    private enum CodingKeys: String, CodingKey {
        case id, routeId, timestamp, transition
        // parameters requires custom handling
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        routeId = try container.decode(String.self, forKey: .routeId)
        timestamp = try container.decode(Date.self, forKey: .timestamp)
        transition = try container.decode(NavigationTransition.self, forKey: .transition)
        parameters = [:] // Simplified for demo
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(id, forKey: .id)
        try container.encode(routeId, forKey: .routeId)
        try container.encode(timestamp, forKey: .timestamp)
        try container.encode(transition, forKey: .transition)
        // parameters encoding simplified for demo
    }
}

/// Navigation transition types
public enum NavigationTransition: String, Codable, CaseIterable {
    case push
    case pop
    case replace
    case reset
    case deepLink
}

/// Navigation state metadata
public struct NavigationStateMetadata: Codable, Equatable, Hashable {
    public let lastSyncTimestamp: Date?
    public let syncAttempts: Int
    public let conflictResolutionStrategy: ConflictResolutionStrategy
    public let customData: [String: String]

    public init(
        lastSyncTimestamp: Date? = nil,
        syncAttempts: Int = 0,
        conflictResolutionStrategy: ConflictResolutionStrategy = .clientWins,
        customData: [String: String] = [:]
    ) {
        self.lastSyncTimestamp = lastSyncTimestamp
        self.syncAttempts = syncAttempts
        self.conflictResolutionStrategy = conflictResolutionStrategy
        self.customData = customData
    }
}

/// Conflict resolution strategies for sync conflicts
public enum ConflictResolutionStrategy: String, Codable, CaseIterable {
    case clientWins
    case serverWins
    case merge
    case prompt
}

/// Navigation state sync request
public struct NavigationStateSyncRequest: Codable {
    public let userId: String
    public let sessionId: String
    public let platform: String
    public let navigationStack: [NavigationStackEntry]
    public let timestamp: Date
    public let syncVersion: Int

    public init(
        userId: String,
        sessionId: String,
        platform: String,
        navigationStack: [NavigationStackEntry],
        timestamp: Date,
        syncVersion: Int
    ) {
        self.userId = userId
        self.sessionId = sessionId
        self.platform = platform
        self.navigationStack = navigationStack
        self.timestamp = timestamp
        self.syncVersion = syncVersion
    }
}

/// Navigation state sync response
public struct NavigationStateSyncResponse: Codable {
    public let success: Bool
    public let syncVersion: Int
    public let conflictsResolved: [String]
    public let timestamp: Date

    public init(
        success: Bool,
        syncVersion: Int,
        conflictsResolved: [String] = [],
        timestamp: Date = Date()
    ) {
        self.success = success
        self.syncVersion = syncVersion
        self.conflictsResolved = conflictsResolved
        self.timestamp = timestamp
    }
}

// MARK: - Default Navigation States

extension NavigationState {

    /// Create default navigation state for main app flow
    public static func defaultState(
        userId: String,
        sessionId: String
    ) -> NavigationState {
        let rootEntry = NavigationStackEntry(
            routeId: "chat",
            parameters: [:],
            timestamp: Date(),
            transition: .reset
        )

        return NavigationState(
            userId: userId,
            sessionId: sessionId,
            platform: "ios",
            navigationStack: [rootEntry],
            currentRoute: "chat",
            previousRoute: nil,
            timestamp: Date(),
            syncVersion: 1,
            metadata: NavigationStateMetadata()
        )
    }

    /// Create navigation state from deep link
    public static func fromDeepLink(
        userId: String,
        sessionId: String,
        deepLinkUrl: String
    ) -> NavigationState? {
        // Parse deep link and create appropriate navigation state
        // This is a simplified implementation
        guard let url = URL(string: deepLinkUrl),
              url.scheme == "tchat" else {
            return nil
        }

        let pathComponents = url.pathComponents.filter { $0 != "/" }
        guard !pathComponents.isEmpty else {
            return defaultState(userId: userId, sessionId: sessionId)
        }

        let routeId = pathComponents.joined(separator: "/")
        let entry = NavigationStackEntry(
            routeId: routeId,
            parameters: [:], // Could extract query parameters
            timestamp: Date(),
            transition: .deepLink
        )

        return NavigationState(
            userId: userId,
            sessionId: sessionId,
            platform: "ios",
            navigationStack: [entry],
            currentRoute: routeId,
            previousRoute: nil,
            timestamp: Date(),
            syncVersion: 1,
            metadata: NavigationStateMetadata()
        )
    }
}

File: apps/mobile/android/app/src/main/java/com/tchat/models/NavigationState.kt
package com.tchat.models

import kotlinx.serialization.Serializable
import kotlinx.serialization.Contextual
import java.util.Date
import java.util.UUID

/**
 * Core navigation state entity for cross-platform state synchronization
 */
@Serializable
data class NavigationState(
    val id: String = UUID.randomUUID().toString(),
    val userId: String,
    val sessionId: String,
    val platform: String = "android",
    val navigationStack: MutableList<NavigationStackEntry>,
    var currentRoute: String? = null,
    var previousRoute: String? = null,
    val timestamp: Long = System.currentTimeMillis(),
    var syncVersion: Int = 1,
    val metadata: NavigationStateMetadata = NavigationStateMetadata()
) {

    /**
     * Get current navigation depth
     */
    val depth: Int
        get() = navigationStack.size

    /**
     * Check if can go back
     */
    val canGoBack: Boolean
        get() = navigationStack.size > 1

    /**
     * Get root route
     */
    val rootRoute: String?
        get() = navigationStack.firstOrNull()?.routeId

    /**
     * Get current route parameters
     */
    val currentParameters: Map<String, Any>
        get() = navigationStack.lastOrNull()?.parameters ?: emptyMap()

    /**
     * Check if state is synchronized
     */
    val isSynchronized: Boolean
        get() {
            val lastSync = metadata.lastSyncTimestamp ?: return false
            return (System.currentTimeMillis() - lastSync) < 30_000 // 30 seconds
        }

    /**
     * Push new route to navigation stack
     */
    fun push(
        routeId: String,
        parameters: Map<String, Any> = emptyMap(),
        timestamp: Long = System.currentTimeMillis()
    ) {
        val entry = NavigationStackEntry(
            routeId = routeId,
            parameters = parameters,
            timestamp = timestamp,
            transition = NavigationTransition.PUSH
        )

        previousRoute = currentRoute
        currentRoute = routeId
        navigationStack.add(entry)
        incrementVersion()
    }

    /**
     * Pop current route from navigation stack
     */
    fun pop() {
        if (navigationStack.size <= 1) return

        val poppedEntry = navigationStack.removeAt(navigationStack.size - 1)
        previousRoute = currentRoute
        currentRoute = navigationStack.lastOrNull()?.routeId

        // Add reverse transition entry for sync
        val reverseEntry = poppedEntry.copy(
            transition = NavigationTransition.POP,
            timestamp = System.currentTimeMillis()
        )

        incrementVersion()
    }

    /**
     * Replace current route
     */
    fun replace(
        routeId: String,
        parameters: Map<String, Any> = emptyMap(),
        timestamp: Long = System.currentTimeMillis()
    ) {
        if (navigationStack.isEmpty()) {
            push(routeId, parameters, timestamp)
            return
        }

        val entry = NavigationStackEntry(
            routeId = routeId,
            parameters = parameters,
            timestamp = timestamp,
            transition = NavigationTransition.REPLACE
        )

        previousRoute = currentRoute
        currentRoute = routeId
        navigationStack[navigationStack.size - 1] = entry
        incrementVersion()
    }

    /**
     * Reset navigation to root
     */
    fun popToRoot() {
        if (navigationStack.size <= 1) return

        val rootEntry = navigationStack.first()
        previousRoute = currentRoute
        currentRoute = rootEntry.routeId
        navigationStack.clear()
        navigationStack.add(rootEntry)
        incrementVersion()
    }

    /**
     * Reset entire navigation state
     */
    fun reset(
        toRoute: String,
        parameters: Map<String, Any> = emptyMap()
    ) {
        val entry = NavigationStackEntry(
            routeId = toRoute,
            parameters = parameters,
            timestamp = System.currentTimeMillis(),
            transition = NavigationTransition.RESET
        )

        previousRoute = currentRoute
        currentRoute = toRoute
        navigationStack.clear()
        navigationStack.add(entry)
        incrementVersion()
    }

    /**
     * Mark state as synchronized
     */
    fun markSynchronized(): NavigationState {
        return copy(
            metadata = metadata.copy(
                lastSyncTimestamp = System.currentTimeMillis()
            )
        )
    }

    /**
     * Increment sync version
     */
    private fun incrementVersion() {
        syncVersion++
    }

    /**
     * Create sync request payload
     */
    fun createSyncRequest(): NavigationStateSyncRequest {
        return NavigationStateSyncRequest(
            userId = userId,
            sessionId = sessionId,
            platform = platform,
            navigationStack = navigationStack.toList(),
            timestamp = System.currentTimeMillis(),
            syncVersion = syncVersion
        )
    }

    /**
     * Apply sync response
     */
    fun applySyncResponse(response: NavigationStateSyncResponse): NavigationState {
        return if (response.success) {
            copy(syncVersion = response.syncVersion).markSynchronized()
        } else {
            this
        }
    }

    companion object {
        /**
         * Create default navigation state for main app flow
         */
        fun defaultState(
            userId: String,
            sessionId: String
        ): NavigationState {
            val rootEntry = NavigationStackEntry(
                routeId = "chat",
                parameters = emptyMap(),
                timestamp = System.currentTimeMillis(),
                transition = NavigationTransition.RESET
            )

            return NavigationState(
                userId = userId,
                sessionId = sessionId,
                platform = "android",
                navigationStack = mutableListOf(rootEntry),
                currentRoute = "chat",
                previousRoute = null,
                timestamp = System.currentTimeMillis(),
                syncVersion = 1,
                metadata = NavigationStateMetadata()
            )
        }

        /**
         * Create navigation state from deep link
         */
        fun fromDeepLink(
            userId: String,
            sessionId: String,
            deepLinkUrl: String
        ): NavigationState? {
            // Parse deep link and create appropriate navigation state
            // This is a simplified implementation
            if (!deepLinkUrl.startsWith("tchat://")) {
                return null
            }

            val path = deepLinkUrl.removePrefix("tchat://").trim('/')
            if (path.isEmpty()) {
                return defaultState(userId, sessionId)
            }

            val routeId = path.replace("/", "/")
            val entry = NavigationStackEntry(
                routeId = routeId,
                parameters = emptyMap(), // Could extract query parameters
                timestamp = System.currentTimeMillis(),
                transition = NavigationTransition.DEEP_LINK
            )

            return NavigationState(
                userId = userId,
                sessionId = sessionId,
                platform = "android",
                navigationStack = mutableListOf(entry),
                currentRoute = routeId,
                previousRoute = null,
                timestamp = System.currentTimeMillis(),
                syncVersion = 1,
                metadata = NavigationStateMetadata()
            )
        }
    }
}

/**
 * Navigation stack entry with transition information
 */
@Serializable
data class NavigationStackEntry(
    val id: String = UUID.randomUUID().toString(),
    val routeId: String,
    val parameters: Map<String, @Contextual Any> = emptyMap(),
    val timestamp: Long = System.currentTimeMillis(),
    var transition: NavigationTransition = NavigationTransition.PUSH
)

/**
 * Navigation transition types
 */
@Serializable
enum class NavigationTransition {
    PUSH,
    POP,
    REPLACE,
    RESET,
    DEEP_LINK
}

/**
 * Navigation state metadata
 */
@Serializable
data class NavigationStateMetadata(
    val lastSyncTimestamp: Long? = null,
    val syncAttempts: Int = 0,
    val conflictResolutionStrategy: ConflictResolutionStrategy = ConflictResolutionStrategy.CLIENT_WINS,
    val customData: Map<String, String> = emptyMap()
)


/**
 * Navigation state sync request
 */
@Serializable
data class NavigationStateSyncRequest(
    val userId: String,
    val sessionId: String,
    val platform: String,
    val navigationStack: List<NavigationStackEntry>,
    val timestamp: Long,
    val syncVersion: Int
)

/**
 * Navigation state sync response
 */
@Serializable
data class NavigationStateSyncResponse(
    val success: Boolean,
    val syncVersion: Int,
    val conflictsResolved: List<String> = emptyList(),
    val timestamp: Long = System.currentTimeMillis()
)